numba/unicode.py

```diff
@@ -1361,6 +1361,43 @@ def impl(a):
     return impl
 
 
+# https://github.com/python/cpython/blob/1d4b6ba19466aba0eb91c4ba01ba509acf18c723/Objects/unicodeobject.c#L9819-L9834    # noqa: E501
+@overload_method(types.UnicodeType, 'casefold')
+def unicode_casefold(data):
+    """Implements str.casefold()"""
+    def impl(data):
+        # https://github.com/python/cpython/blob/1d4b6ba19466aba0eb91c4ba01ba509acf18c723/Objects/unicodeobject.c#L10782-L10791    # noqa: E501
+        def _do_casefold(data, length, res, maxchars):
+            """Translation of the function to case fold a unicode string."""
+            k = 0
+            mapped = np.zeros(3, dtype=_Py_UCS4)
+            for idx in range(length):
+                mapped.fill(0)
+                code_point = _get_code_point(data, idx)
+                n_res = _PyUnicode_ToFoldedFull(code_point, mapped)
+                for m in mapped[:n_res]:
+                    maxchar = maxchars[0]
+                    maxchars[0] = max(maxchar, m)
+                    _set_code_point(res, k, m)
+                    k += 1
+            return k
+
+        length = len(data)
+        tmp = _empty_string(PY_UNICODE_4BYTE_KIND, 3 * length, data._is_ascii)
+        # maxchar should be inside of a list to be pass as argument by reference
+        maxchars = [0]
+        newlength = _do_casefold(data, length, tmp, maxchars)
+        maxchar = maxchars[0]
+        newkind = _codepoint_to_kind(maxchar)
+        res = _empty_string(newkind, newlength, _codepoint_is_ascii(maxchar))
+        for i in range(newlength):
+            _set_code_point(res, i, _get_code_point(tmp, i))
+
+        return res
```

Comment:
Same as https://github.com/numba/numba/pull/4823/files#diff-a21787256ad779b4c4c5c99915c71025R1301-R1311, perhaps this should be a function closing over the `_do_XYZ` impl?!

Reply:
Merged code of `_do_casefold()` to the main code.

Reply:
I'm not sure how this change improves the situation, the closure that was present previously would have just been inlined at no cost?

The issue I was alluding to here is that the section of code above is common to a lot of the unicode implementations, we could just make it a function. e.g.

```python
def generate_loop_apply_unicode_func(FUNCTION):
    def impl(<args>):
            k = 0
            maxchar = 0
            mapped = np.zeros(3, dtype=_Py_UCS4)
            tmp = _empty_string(PY_UNICODE_4BYTE_KIND, 3 * length)
            for idx in range(length):
                mapped.fill(0)
                code_point = _get_code_point(data, idx)
                n_res = FUNCTION(code_point, mapped)
                for m in mapped[:n_res]:
                    maxchar = max(maxchar, m)
                    _set_code_point(tmp, k, m)
                    k += 1
            newkind = _codepoint_to_kind(maxchar)
            res = _empty_string(newkind, k)
            for i in range(k):
                _set_code_point(res, i, _get_code_point(tmp, i))
    return impl


_apply_ToFoldedFull = register_jitable(generate_loop_apply_unicode_func(_PyUnicode_ToFoldedFull))
```
then use `_apply_ToFoldedFull` in here. It's then easy to use again when the same pattern occurs/is amenable for further generalisation on the basis that all these algorithms look pretty similar.

Reply:
This is something that might help, it can and should be generalised further:
https://github.com/stuartarchibald/numba/blob/4a768e63124b837fdbfb24b69d6b87385733746c/numba/unicode.py#L1312-L1373

Reply:
@stuartarchibald I tried to apply your recommendation. Could you please recheck the code?

Reply:
Thanks for trying that out. What did you think? Can you apply what you've done to more cases without modification to massive reduce the implementation size? If it's not possible to do this then the abstraction might not be quite right yet.

Reply:
I can apply that at the least for methods `str.capitalize()` and `str.swapcase()` after the PR changes are in master.

Reply:
> I can apply that at the least for methods `str.capitalize()` and `str.swapcase()` after the PR changes are in master.

Sounds good, look forward to seeing it :)

Is there any reason not to include https://github.com/numba/numba/pull/4824/files#diff-a21787256ad779b4c4c5c99915c71025R1407-R1416 in the abstraction too?

Reply:
@stuartarchibald sorry, but what do you mean under the abstraction? Maybe you could provide some example to make me understand what you mean.

Reply:
@stuartarchibald If you mean to move https://github.com/numba/numba/pull/4824/files#diff-a21787256ad779b4c4c5c99915c71025R1407-R1416 to `generate_unicode_operation_doer`, in this case I won't be able to reuse it in `str.capitalize()`. I think I will add separate doer instead of `generate_unicode_operation_doer` for the method.

Reply:
@stuartarchibald maybe we will move the PR on to be able to reuse the code generation in `str.capitalize()` and `str.swapcase()`?

Reply:
> @stuartarchibald If you mean to move https://github.com/numba/numba/pull/4824/files#diff-a21787256ad779b4c4c5c99915c71025R1407-R1416 to `generate_unicode_operation_doer`, in this case I won't be able to reuse it in `str.capitalize()`. I think I will add separate doer instead of `generate_unicode_operation_doer` for the method.

Right, so this is my concern about the level of generalisation available in the current implementation. Adding a specific generator for each implementation somewhat defeats the point of having generators?

Reply:
Anyway [`generate_common_operation_doer()`](https://github.com/numba/numba/pull/4824/files#diff-a21787256ad779b4c4c5c99915c71025R1394) is going to be common generator, but `generate_unicode_operation_doer` and `generate_ascii_operation_doer` are unlikely to be be reused in another methods..

Reply:
It was a reason why I separated `generate_common_operation_doer()` and `generate_unicode_operation_doer()`+`generate_ascii_operation_doer()`.

Reply:
@stuartarchibald I changed generating of case operation methods leaving only `generate_common_operation_doer()` -> `generate_operation_func()` as common generator which can be reused for other methods (`str.capitalize`, `str.swapcase`). So could you please re-review the change?